pkg/controller/vsphere/reconciler.go

```diff
@@ -231,18 +304,94 @@ func clone(s *machineScope) error {
 	}
 
-	// TODO: store task in providerStatus/conditions?
-	klog.V(3).Infof("%v: running task: %v", s.machine.GetName(), task.Name())
+	s.providerStatus.TaskRef = task.Reference().Value
+	klog.V(3).Infof("%v: running task: %+v", s.machine.GetName(), s.providerStatus.TaskRef)
 	return nil
 }
 
+func getNetworkDevices(s *machineScope, devices object.VirtualDeviceList) ([]types.BaseVirtualDeviceConfigSpec, error) {
+	var networkDevices []types.BaseVirtualDeviceConfigSpec
+	// Remove any existing NICs
+	for _, dev := range devices.SelectByType((*types.VirtualEthernetCard)(nil)) {
+		networkDevices = append(networkDevices, &types.VirtualDeviceConfigSpec{
+			Device:    dev,
+			Operation: types.VirtualDeviceConfigSpecOperationRemove,
+		})
+	}
+
+	// Add new NICs based on the machine config.
+	key := int32(-100)
+	for i := range s.providerSpec.Network.Devices {
+		netSpec := &s.providerSpec.Network.Devices[i]
+		klog.V(3).Infof("Adding device: %v", netSpec.NetworkName)
+
+		ref, err := s.GetSession().Finder.Network(s.Context, netSpec.NetworkName)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to find network %q", netSpec.NetworkName)
+		}
+
+		backing, err := ref.EthernetCardBackingInfo(s.Context)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to create new ethernet card backing info for network %q", netSpec.NetworkName)
+		}
+
+		dev, err := object.EthernetCardTypes().CreateEthernetCard(ethCardType, backing)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to create new ethernet card %q for network %q", ethCardType, netSpec.NetworkName)
+		}
+
+		// Get the actual NIC object. This is safe to assert without a check
+		// because "object.EthernetCardTypes().CreateEthernetCard" returns a
+		// "types.BaseVirtualEthernetCard" as a "types.BaseVirtualDevice".
+		nic := dev.(types.BaseVirtualEthernetCard).GetVirtualEthernetCard()
+		// Assign a temporary device key to ensure that a unique one will be
+		// generated when the device is created.
+		nic.Key = key
+
+		networkDevices = append(networkDevices, &types.VirtualDeviceConfigSpec{
+			Device:    dev,
+			Operation: types.VirtualDeviceConfigSpecOperationAdd,
+		})
+		klog.V(3).Infof("Adding device: eth card type: %v, network spec: %+v, device info: %+v",
+			ethCardType, netSpec, dev.GetVirtualDevice().Backing)
+		key--
+	}
+
+	return networkDevices, nil
+}
+
 func newVMFlagInfo() *types.VirtualMachineFlagInfo {
 	diskUUIDEnabled := true
 	return &types.VirtualMachineFlagInfo{
 		DiskUuidEnabled: &diskUUIDEnabled,
 	}
 }
 
+func taskIsFinished(task *mo.Task) (bool, error) {
+	if task == nil {
+		return true, nil
+	}
+
+	// Otherwise the course of action is determined by the state of the task.
+	klog.V(3).Infof("task: %v, state: %v, description-id: %v", task.Reference().Value, task.Info.State, task.Info.DescriptionId)
+	switch task.Info.State {
+	case types.TaskInfoStateQueued:
+		klog.V(3).Info("task is still pending")
```

Comment:
I guess it doesn't matter too much, but this log line seems redundant given the one before the switch statement. We should probably at least add the task ID / name / whatever here if this is necessary.